### PR TITLE
improved field selection

### PR DESCRIPTION
**Pull request by jgraettinger**
**Description:**

This PR implements the improved field selection system for materializations, introducing support for custom group-by keys, depth-based field selection, and centralizing field validation logic in the control plane. The implementation runs in "shadow mode" alongside existing field selection logic to enable safe validation and analysis of changes before full cutover.

Key enhancements include:
- **Custom Group-by Keys**: Users can now specify `groupBy` fields independent of collection key structure
- **Depth-based Selection**: Control materialization depth to avoid over-wide tables from deeply nested documents  
- **Centralized Validation**: Field constraints and folding logic moved from individual connectors to control plane
- **Intelligent Field Exclusion**: Automatically excludes nested fields when parent fields are selected
- **Shadow Mode Implementation**: New logic runs parallel to existing system for safe validation

**Workflow steps:**

The feature introduces several new configuration options while maintaining backward compatibility:

1. **Custom Primary Keys**: In materialization bindings, users can now specify a `groupBy` array to define custom primary keys:
   ```yaml
   bindings:
     - resource: { table: my_table }
       source: my/collection
       groupBy:
         - /user_id
         - /timestamp
   ```

2. **Depth Control**: The `recommended` field now accepts integer depth values:
   ```yaml
   fields:
     recommended: 2  # Select fields up to depth 2
     require: ["/important_field"]
     exclude: ["/sensitive_data"]
   ```

3. **Enhanced Field Selection**: The control plane now provides detailed per-field selection rationale and conflict detection, accessible via new WASM functions for UI integration.

Current workflow remains unchanged - existing materializations continue to work with their current field selections preserved.

**Documentation links affected:**

- [Improved Field Selection Plan](../blob/johnny/field-selection/plans/improved-field-selection.md) - Implementation plan with detailed technical specifications
- [Flow Web WASM Field Selection](../blob/johnny/field-selection/crates/flow-web/FIELD_SELECTION.md) - New documentation for field selection evaluation API
- [Flow Web WASM Collection Projections](../blob/johnny/field-selection/crates/flow-web/COLLECTION_PROJECTIONS.md) - New documentation for collection validation API  
- _Primary documentation has not been updated yet_

**Notes for reviewers:**

**Shadow Mode Strategy**: This implementation deliberately runs the new field selection logic in parallel with existing logic, logging differences without affecting production materializations. This allows us to:
- Validate the implementation against real-world scenarios
- Identify materializations that would be impacted by collection key handling changes
- Provide a safe migration path for existing tasks

**Key Technical Changes**:
- `MaterializationFields` model extended with `groupBy` and depth-based `recommended` field
- New `Select`/`Reject` enum system for priority-based field selection decisions  
- Enhanced materialization constraint protocol with `folded_field` support
- Comprehensive snapshot testing with `insta` for field selection scenarios
- WASM bindings for UI integration of new field selection logic

**Backward Compatibility**: 
- Existing `include`/`exclude`/`recommended` configurations continue to work
- Field selections for running materializations are preserved unless explicitly changed
- Collection key handling changes are the primary breaking change concern, mitigated through shadow mode analysis

**Testing Coverage**:
- 10+ snapshot tests covering various field selection scenarios and edge cases
- Validation against ops-catalog workflows to ensure production compatibility
- WASM function testing for UI integration points

The implementation represents Phase 1 and Phase 2 of the planned rollout. Phase 3 (cutover to new logic) will follow after shadow mode validation period.

- - -
This change is [Reviewable](https://reviewable.io/reviews/estuary/flow/2195)

### CODE REVIEW

**Comment by jgraettinger** [comment]
The first commit is a "plan", a new experiment I'm trying for agentic development, which also produced a nice documentation artifact of the feature & goals.

@williamhbaker, would you review commits related to materialization protocol changes and the implementation of field selection logic?

@kiahna-tucker, would you review the commit which integrates with flow-web, particularly the interfaces / docs of how it works / inputs / outputs?

Thanks. Best to review this commit by commit

File: plans/improved-field-selection.md [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+**2. Refined Constraint Protocol**
+
+Materialization constraints will be clarified with updated semantics and a new `folded_field` capability:
+- `FIELD_REQUIRED` - Fields with special connector meaning (e.g., "text" field for Slack connector)
+- `LOCATION_REQUIRED` - JSON pointer locations that must be present (e.g., root document, all top-level fields)
+- `FIELD_OPTIONAL` - Fields that may be materialized (absorbs deprecated `LOCATION_RECOMMENDED`)
+- `FIELD_FORBIDDEN` - Fields permanently unusable (e.g., unsupported types like JSON null)
+- `UNSATISFIABLE` - Fields requiring backfill to resolve conflicts (e.g., incompatible SQL column types)
+- `folded_field` - Connector-provided field name transformations for technical limitations
```
Comment (kiahna-tucker) [inline plans/improved-field-selection.md:68]: Will all the characters of this constraint type be lowercase (as written here)? And, out of curiosity, why use the term `folded_field` as opposed to `field_folded`? The latter aligns with the running naming convention (i.e., `<target>_<specifier>`).

File: crates/flow-web/FIELD_SELECTION.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+- **GroupByKey**: Part of the materialization group-by key
+- **CurrentDocument**: Currently used to store the document
+- **UserRequires**: Required by user's field selection
+- **ConnectorRequires**: Required by connector
+- **PartitionKey**: Collection partition key
+- **CurrentValue**: Part of current materialization
+- **UserDefined**: User-projected field
+- **ConnectorRequiresLocation**: Location-based connector requirement
+- **CoreMetadata**: Essential system fields
+- **DesiredDepth**: Within user-specified depth
```
Comment (kiahna-tucker) [inline crates/flow-web/FIELD_SELECTION.md:73]: Which of these selection reasons indicate that the field is _required_? Or will this function only determine whether a field is _selected_ or _excluded_? Should constraint types be leveraged to determine whether a _selected_ field is _required_?

File: plans/improved-field-selection.md (hunk 1 above)
Comment (jgraettinger) [reply plans/improved-field-selection.md:68]: this is mixing constraints given as protobuf constants (e.x. `FIELD_REQUIRED`), with a new protobuf message field `folded_field: string`.

They're all "constraints" in the formal sense but mean different things.
The UI doesn't see these raw protobuf constants, only their encodings as Select / Reject reasons

File: crates/flow-web/FIELD_SELECTION.md (hunk 2 above)
Comment (jgraettinger) [reply crates/flow-web/FIELD_SELECTION.md:73]: There are multiple reasons why a field can be "required": GroupByKey, UserRequires -- which is the `require` portion of the model; ConnectorRequires, ConnectorRequiresLocation

This routine is giving you the final outcome of field selection: whether it will be selected or not.

I _don't_ envision the UI using the structured `reason` fields on `SelectOutput` / `RejectOutput`, at least not at first. I **am** expecting that the `outcome.select.reason` and `outcome.reject.reason` would be displayed, though, perhaps as a hover or something.

**Review by williamhbaker — APPROVED** [review]
LGTM
